code/datums/components/shielded.dm

```diff
@@ -0,0 +1,170 @@
+	var/datum/callback/on_hit_effects
+
+/datum/component/shielded/Initialize(max_charges = 3, recharge_start_delay = 20 SECONDS, charge_increment_delay = 1 SECONDS, shield_icon_file = 'icons/effects/effects.dmi', shield_icon = "shield-old", shield_inhand = FALSE, run_hit_callback)
+	if(!isitem(parent) || max_charges <= 0)
+		return COMPONENT_INCOMPATIBLE
+
+	src.max_charges = max_charges
+	src.recharge_start_delay = recharge_start_delay
+	src.charge_increment_delay = charge_increment_delay
+	src.shield_icon_file = shield_icon_file
+	src.shield_icon = shield_icon
+	src.shield_inhand = shield_inhand
+	src.on_hit_effects = run_hit_callback || CALLBACK(src, .proc/default_run_hit_callback)
+
+	current_charges = max_charges
+	if(recharge_start_delay)
+		START_PROCESSING(SSdcs, src)
+
+/datum/component/shielded/Destroy(force, silent)
+	if(wearer)
+		shield_icon = "broken"
+		UnregisterSignal(wearer, COMSIG_ATOM_UPDATE_OVERLAYS)
+		wearer.update_appearance(UPDATE_ICON)
+		wearer = null
+	QDEL_NULL(on_hit_effects)
+	return ..()
+
+/datum/component/shielded/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/on_equipped)
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, .proc/on_dropped)
+	RegisterSignal(parent, COMSIG_ITEM_HIT_REACT, .proc/on_hit_react)
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/check_recharge_rune)
+
+/datum/component/shielded/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED, COMSIG_ITEM_HIT_REACT, COMSIG_PARENT_ATTACKBY))
+
+// Handle recharging, if we want to
+/datum/component/shielded/process(delta_time)
+	if(current_charges >= max_charges)
+		STOP_PROCESSING(SSdcs, src)
+		return
+
+	if(!COOLDOWN_FINISHED(src, recently_hit_cd))
+		return
+	if(!COOLDOWN_FINISHED(src, charge_add_cd))
+		return
+
+	var/obj/item/item_parent = parent
+	COOLDOWN_START(src, charge_add_cd, charge_increment_delay)
+	current_charges++
+	if(wearer && current_charges == 1)
+		wearer.update_appearance(UPDATE_ICON)
+	playsound(item_parent.loc, 'sound/magic/charge.ogg', 50, TRUE)
+	if(current_charges == max_charges)
+		playsound(item_parent.loc, 'sound/machines/ding.ogg', 50, TRUE)
+
+/// Check if we've been equipped to a valid slot to shield
+/datum/component/shielded/proc/on_equipped(datum/source, mob/user, slot)
+	SIGNAL_HANDLER
+
+	if(slot == ITEM_SLOT_HANDS && !shield_inhand)
+		on_dropped(source, user)
+		return
+
+	wearer = user
+	RegisterSignal(wearer, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/on_update_overlays)
+	RegisterSignal(wearer, COMSIG_PARENT_QDELETING, .proc/on_wearer_qdel)
+	if(current_charges)
+		wearer.update_appearance(UPDATE_ICON)
+
+/// When dropped, forget about whoever we were worn by
+/datum/component/shielded/proc/on_dropped(datum/source, mob/user)
+	SIGNAL_HANDLER
+
+	if(wearer)
+		UnregisterSignal(wearer, list(COMSIG_ATOM_UPDATE_OVERLAYS, COMSIG_PARENT_QDELETING))
+		wearer.update_appearance(UPDATE_ICON)
+		wearer = null
```

Comment:
This snippet is repeated in `on_wearer_qdel` and `Destroy`. Can you move it into it's own proc?